source-code/core/src/ast/schema.ts

```diff
@@ -1,59 +1,77 @@
+export type NodeName =
+  | 'Identifier'
+  | 'Resource'
+  | 'Message'
+  | 'Pattern'
+  | 'Text'
+  | 'LanguageTag'
+
+type ExtensionInformation = {
+  [node in NodeName | 'Node']?: Record<string, unknown>
+}
+
 /**
  * A single node of the AST.
  *
- * Every other defintions are based on Node.
+ * Every other definitions are based on Node.
  */
-export type Node<Name> = {
-  type: Name;
-  /**
-   * Metadata is ignored by inlang.
-   *
-   * Use the metadata property to store additional
-   * information for a particular node like parsing
-   * and serialization information.
-   */
-  metadata?: any;
-};
+type Node<Name extends NodeName, Extension extends ExtensionInformation = ExtensionInformation> =
+  Extension[Name] & Extension['Node'] & {
+    type: Name
+    /**
+     * Metadata is ignored by inlang.
+     *
+     * Use the metadata property to store additional
+     * information for a particular node like parsing
+     * and serialization information.
+     */
+    metadata?: any;
+  }
 
 /**
  * An identifier.
  *
  * Some Nodes have Identifiers such as a Resource or Message.
  */
-export type Identifier = Node<"Identifier"> & {
-  name: string;
-};
+export type Identifier<Extension extends ExtensionInformation = ExtensionInformation> =
+  Node<"Identifier", Extension> & {
+    name: string;
+  };
 
 /**
  * A resource is a collection of messages.
  */
-export type Resource = Node<"Resource"> & {
-  languageTag: LanguageTag;
-  body: Array<Message>;
-};
+export type Resource<Extension extends ExtensionInformation = ExtensionInformation> =
+  Node<"Resource", Extension> & {
+    languageTag: LanguageTag<Extension>;
+    body: Array<Message<Extension>>;
```

Comment:
I saw a utility type on Twitter that could improve the readability of the Nodes with extensions on hovering.

Reply:
I was thinking about the same and tried it already. Didn't work and made the type more complex.

Reply:
Sad to hear :(